### DISCUSSION TITLE
Making dependancies leaves

### DISCUSSION THREAD

**Discussion by johnalanwoods** [Homebrew/discussions#5821] [Getting started] [answered]
### Output of `brew config`

```shell
: 4.4.12
ORIGIN: https://github.com/Homebrew/brew
HEAD: 3001afe88112c13b5aafdd92e097680897060881
Last commit: 6 days ago
Branch: stable
Core tap JSON: 21 Dec 16:28 UTC
Core cask tap JSON: 21 Dec 16:28 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 3.3.6 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 16-core 64-bit arm_brava
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.2-arm64
CLT: 16.2.0.0.1.1733547573
Xcode: 16.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
`Your system is ready to brew.`
```

### Description of issue

Hello!

I’m using Homebrew on macOS and would like to make sure that certain formulas, like Python, remain installed even if they were initially installed as a dependency of another package. 

Many critical packages get installed as deps for other trivial ones.

Is there a recommended best practice for promoting certain tools to "leaves" or having them recognised as “top-level” so that uninstalling a dependent package does not inadvertently remove them? 

I’ve seen references to brew pin, but I’m curious if there are other approaches or if pinning is indeed the suggested method in this situation - doesn't feel like it as it would avoid upgrades.

Essentially if not I guess the order of install is very important, from critical to less critical?

Thanks for your help, and for all the work on this fantastic project!

Enjoy Christmas.

**Comment by carlocab** [comment, marked as answer]
You probably want `brew tab`.
```
❯ brew help tab
Usage: brew tab [options] installed_formula|installed_cask [...]

Edit tab information for installed formulae or casks.

This can be useful when you want to control whether an installed formula should
be removed by brew autoremove. To prevent removal, mark the formula as
installed on request; to allow removal, mark the formula as not installed on
request.

      --installed-on-request       Mark installed_formula or installed_cask
                                   as installed on request.
      --no-installed-on-request    Mark installed_formula or installed_cask
                                   as not installed on request.
      --formula, --formulae        Only mark formulae.
      --cask, --casks              Only mark casks.
  -d, --debug                      Display any debugging information.
  -q, --quiet                      Make some output more quiet.
  -v, --verbose                    Make some output more verbose.
  -h, --help                       Show this message.
```

**Reply by johnalanwoods** [reply]
thank you and regards @carlocab - I know you contribute a lot on this project, just like @SMillerDev.

So does `--installed-on-request` essentially promote formula to show in brew leaves - testing it doesn't seem to, maybe as it still has a dependent.

Otherwise I guess the right approach is appropriate install order?

**Reply by carlocab** [reply]
> So does `--installed-on-request` essentially promote formula to show in brew leaves - testing it doesn't seem to, maybe as it still has a dependent.

Not quite. It just tells `brew` to consider the formula as installed-on-request so it doesn't get automatically removed by `brew autoremove` (whether invoked manually, or automatically by, e.g., `brew cleanup`).

`brew leaves` lists formulae that have no dependents:
```
❯ brew help leaves
Usage: brew leaves [--installed-on-request] [--installed-as-dependency]

List installed formulae that are not dependencies of another installed formula
or cask.
[snip]
```
So, if you want a formula to show up in `brew leaves`, you must uninstall all its dependents. You should be able to do that with something like
```
brew uninstall "$(brew uses --installed $formula)"
```

**Reply by johnalanwoods** [reply]
Many thanks!
